### ISSUE TITLE
Haddock to LaTeX fails with internal error

### ISSUE THREAD

**Issue by pjljvandelaar** [haskell/haddock#930]
When I run haddock as follows on the `valexpr` package of TorXakis (https://github.com/TorXakis/TorXakis):
```
C:\> stack haddock .\sys\valexpr --haddock-arguments "--latex" --no-haddock-deps --no-haddock-hyperlink-source
```
I get
```
haddock: internal error: declaration not supported by declNames
CallStack (from HasCallStack):
  error, called at utils\haddock\haddock-api\src\Haddock\Backends\LaTeX.hs:257:8 in main:Haddock.Backends.LaTeX
Completed 5 action(s).
```

`declNames` is not defined within TorXakis, so I have no clue what is causing this error: is the haddock documentation wrong or is it a bug in haddock?

**Comment by harpocrates** [comment]
This is a bug in the LaTeX backend. Any `haddock: internal error` gotten after running `haddock` from the command line indicates some sort of bug.

There should be a bunch of LaTeX backend bugs fixed in the upcoming GHC-8.6.1 release, some of which might be what you are running into (are you using type families, for instance? https://github.com/haskell/haddock/pull/734).

That said, do you think you could try to reduce this down to a test case with a simpler build process? As it stands, I don't have time to reduce (and I can't debug on the project as is, because it is only set up to build with `stack`).

**Comment by pjljvandelaar** [comment]
I don't think we are using type families, but I will check.

Since nothing within TorXakis is mentioned, I have no clue how to reduce it (except by randomly cutting it into pieces).
When GHC-8.6.1 becomes available, I can check whether the bug is gone.

**Comment by bfrk** [comment]
I have just been running into a similar problem with the haddock that comes with ghc-8.6.3. I get the not too helpful error message

```
> cabal new-haddock --haddock-options '--latex'
...
haddock: internal error: LaTeX.ppClassDecl
CallStack (from HasCallStack):
  error, called at utils/haddock/haddock-api/src/Haddock/Backends/LaTeX.hs:608:21 in main:Haddock.Backends.LaTeX
```

I would gladly help by cutting this down to a small example but the error message doesn't give me clue how to do that. This is a large project with over 200 modules.

**Comment by harpocrates** [comment]
@bfrk Venturing a guess: based on the error message, it looks like you have a class with associated types. This isn't something Haddock supports yet in the LaTeX backend.

https://github.com/haskell/haddock/blob/53997f3db71d113bdad59548e3f16adfe90c112b/haddock-api/src/Haddock/Backends/LaTeX.hs#L602-L606

I know there's a fix laying around somewhere. Let me see if I can find it so we can unbreak this.

**Comment by bfrk** [comment]
Yes, I have classes with associated types. Many thanks for the various fixes. I am compiling haddock at this moment... will report back when I know my particular problem is fixed or not.

**Comment by harpocrates** [comment]
> Yes, I have classes with associated types. Many thanks for the various fixes. I am compiling haddock at this moment... will report back when I know my particular problem is fixed or not.

Note that the fix I merged in is currently only in the 8.8 branch, so you need GHC 8.8 to build it...

**Comment by bfrk** [comment]
Um. I don't have a ghc-8.8 available. Can you, ...what's the term in git... cherry-pick? the changes to the ghc-8.6 branch?

**Comment by harpocrates** [comment]
Unfortunately, that’s more work than it's worth.

Given that the 8.8.1 alpha release is imminent, you should be able to try this out soon anyways. :)

**Comment by bfrk** [comment]
I see. That's why I prefer darcs over git: doing things like that is usually no work at all.

**Comment by harpocrates** [comment]
It’s not a problem of finding the right git commands, it’s just an issue of resolving the conflicts that come from re-applying changes on branches that have diverged. If you feel yourself up to it, you can do this yourself. Here are the git incantations you’ll need:

```
git checkout ghc-8.6
git cherry-pick abb448ff120d6f09b6d070806de1d0eb334bc23b # Better support for default methods in classes
git cherry-pick ae23b4f25a972620686617b5aab5375d5046b1c9 # Many LaTeX backend fixes
```

Resolving conflicts (after each cherry pick) is not going to be trivial though (and darcs wouldn’t help you do this either). You may find it easier to just build ghc-8.8 yourself.

**Comment by bfrk** [comment]
You are right. Didn't expect the branches diverged so much. The first patch is simple (only conflict in CHANGES.md) but the second looks ... difficult. I may try it but don't hold your breath.

**Comment by harpocrates** [comment]
Fixed in ae23b4f25a972620686617b5aab5375d5046b1c9.
